### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by rgurzhiy** [BoltsFramework/Bolts-ObjC#249]
I've made a simple OSX application:

**AppDelegate.m**
```
#import "AppDelegate.h"
#import <Bolts/Bolts.h>

@interface AppDelegate ()
@property (weak) IBOutlet NSWindow *window;
@end

@implementation AppDelegate

- (IBAction) test:(id)sender {
    [[self testAsync] continueWithBlock:^id (BFTask *task) {
        NSLog(@"%@", task.result);
        return nil;
    }];
}

- (BFTask *) testAsync {
    BFTaskCompletionSource *completionSource = [BFTaskCompletionSource taskCompletionSource];
    
    [[BFTask taskWithDelay:1000] continueWithBlock:^id (BFTask *task) {
        [completionSource setResult:[NSDate date]];
        return nil;
    }];
    
    return completionSource.task;
}

@end
```
When I run it with Allocations instrument, I get the trace 
[BoltsMemory.trace.zip](https://github.com/BoltsFramework/Bolts-ObjC/files/272374/BoltsMemory.trace.zip)

Almost every time `- (IBAction) test:(id)sender ` is being executed new allocations are added to the trace and stay `Live` forever.
![screen shot 2016-05-19 at 6 48 32 pm](https://cloud.githubusercontent.com/assets/16662707/15391200/715d350e-1df2-11e6-997a-7614523f2df9.png)
As you can see responsible caller is `-[BFTask trySetResult:]`
![screen shot 2016-05-19 at 6 51 57 pm](https://cloud.githubusercontent.com/assets/16662707/15391264/ccf16b42-1df2-11e6-867a-2bfb1eb2bf56.png)
or `-[BFTask continueWithExecutor:block:cancellationToken:]`
![screen shot 2016-05-19 at 6 52 40 pm](https://cloud.githubusercontent.com/assets/16662707/15391295/e855180c-1df2-11e6-904c-4deb971ee38a.png)

Can it be considered as memory leak?

**Comment by nlutsenko** [comment]
Memory leak - maybe...
Most likely - not cleaned up heap allocation.
@rgurzhiy are you profiling in Debug or Release mode?
If you swap them - does it make any difference?

Will gladly investigate deeper, but don't have a time frame just yet.

**Comment by rgurzhiy** [comment]
No difference in Release mode.

Actually it seems like I've found the answer.
This code shows the same result:
```
- (IBAction) test:(id)sender {
    NSObject *lock = [NSObject new];
    @synchronized (lock) {
        NSLog(@"test");
}
```

Stack trace in Instruments looks like this:
![screen shot 2016-05-20 at 2 53 53 pm](https://cloud.githubusercontent.com/assets/16662707/15419729/bbc818cc-1e9a-11e6-97e2-d3e4336501f4.png)

If you look closer at `id2data` function [here](https://github.com/opensource-apple/objc4/blob/master/runtime/objc-sync.mm) you can see that it uses SyncCache.
Since the code above is hell-simple and shows the same result, it seems like this allocations are responsibility of runtime and it should take care of them at some point of execution. 
But should it? Anyway, the issue is hardly possible to deal with Bolts. 

The only thing we can do is using `NSLock` instead of  `@synchronized`. This code doesn't produce `Live` allocations:
```
- (IBAction) test:(id)sender {
    NSLock *lock = [NSLock new];
    [lock lock];
    NSLog(@"test");
    [lock unlock];
}
```

**Comment by richardjrossiii** [comment]
For posterity's sake, here's the relevant section in [objc-sync.mm](http://opensource.apple.com/source/objc4/objc4-680/runtime/objc-sync.mm)

    // malloc a new SyncData and add to list.
    // XXX calling malloc with a global lock held is bad practice,
    // might be worth releasing the lock, mallocing, and searching again.
    // But since we never free these guys we won't be stuck in malloc very often.
    result = (SyncData*)calloc(sizeof(SyncData), 1);
    result->object = (objc_object *)object;
    result->threadCount = 1;
    new (&result->mutex) recursive_mutex_t();
    result->nextData = *listp;
    *listp = result;

It would appear that this is not a bug, but an actual intended behavior for the behavior of @synchronized.

Eventually, switching to `NSRecursiveLock` or similar would be a good idea, or even a direct pthead_mutex_t.

**Comment by rgurzhiy** [comment]
Yep. This is exactly what I've discovered about `@synchronized`.

On the other hand, don't you think that switching to 'manual' locking will make code more complicated?
Actually, I've already refactored `BFTask` to use `NSRecursiveLock`. During work I've got an annoying issue. I've missed one `return` statement between `lock` and `unlock` calls and it took me a while to find out what the hell is going on. I mean, after switching to 'manual' locking, during future refactoring you have to keep in mind this lock-unlock tricky game. With `@synchronized` code looks much cleaner.
If this still sounds like a good idea for you, I can try to get rid of `@synchronized` in the whole project, add 'Unlocked' tests for methods with locking to avoid sudden `return`s and make a pull request.

I don't understand a big difference between `NSRecursiveLock` and recursive `pthead_mutex_t`. Maybe perfomance? Anyway, switching to direct `pthead_mutex_t` shouldn't be a big problem since code pattern is the same with `NSRecursiveLock `. The only thing we have to keep in mind is destroying mutex in `-dealloc`.

**Comment by richardjrossiii** [comment]
@rgurzhiy `NSRecursiveLock`'s implementation under the hood does just use a `pthread_mutex_t` with a mutex attribute of `PTHREAD_MUTEX_RECURSIVE` (verified by disassembling `Foundation.framework` multiple times).

And yes, the manual locking code is what makes using members of the `NSLocking` protocol hard to use - it's possible to wrap it in a block helper, that will automatically lock/unlock on demand for you, but even that has pitfalls (ObjC generics make it impossible to return a value from a block, so wrapping everything in NSValue/NSNumber becomes tedious).

We've actually discussed this before - as the performance of `@synchronized` is abysmal, and were wondering if we could come up with an architecture more closely based on [.NET Core](https://github.com/dotnet/core)'s TPL implementation (which doesn't use recursive locks).

Of course, [this implementation](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs) is *significantly* more complex than what we currently have in bolts, and while the performance gains might be significant, so would the effort to implement such an architecture.
